### PR TITLE
Debug brunch

### PR DESCRIPTION
**Pull request by SilverSulfide**
I WILL RESOLVE THE MERGE CONFLICTS.

Edit: I resolved the conflicts, calmly.

### CODE REVIEW

**Comment by SilverSulfide** [comment]
Conflicts in neox_args.py and learning_rates.py can be ignored

File: megatron/data/gpt2_dataset.py [hunk 1]
```diff
@@ -155,16 +155,18 @@ def __getitem__(self, idx):
...
+                        if n == rw_indx:
+                            rw = dataset.get(self.doc_idx[doc_index_l])
+                            sample_list.append(
+                                np.array([rw[0] for _ in range(sample_lengths.pop(0))])
+                            )
+                        else:
+                            sample_list.append(
+                                dataset.get(self.doc_idx[doc_index_l], length=offset_l + 1)
+                            )
+                            sample_lengths.append(len(sample_list[-1]))
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: Why'd you do this?

File: megatron/data/gpt2_dataset.py [hunk 2]
```diff
@@ -358,9 +375,9 @@ def _build_index_mappings(
                             doc_i = (doc_i + 1) % len(documents)
                             continue
                     # Just in case we have bad data in the loop...
-                    if np.all(label_dataset.get(doc_i)[:seq_length] == -100):
-                        doc_i = (doc_i + 1) % len(documents)
-                        continue
+                    # if np.all(label_dataset.get(doc_i)[:seq_length] == -100):
+                    #     doc_i = (doc_i + 1) % len(documents)
+                    #     continue
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: Random modification to normal packing because...?

File: megatron/data/gpt2_dataset.py [hunk 3]
```diff
@@ -155,16 +155,18 @@ def __getitem__(self, idx):
                             sample_list.append(dataset.get(self.doc_idx[i]))
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: Could we maybe add a docstring to GPT2Dataset like "This is used for getting seq_len size samples of data [batch, seq_len] optionally with rewards, reference logits, and labels"?

File: megatron/data/gpt2_dataset.py [hunk 4]
```diff
@@ -338,7 +342,20 @@ def _build_index_mappings(
...
+                generated_samples = len(sample_idx) - 1  # last row is sentinel-like
+                generated_tokens_cap = generated_samples * (seq_length + 1)
+                true_tokens = sum(sample_lengths)
+                packing_eff_pct = 100.0 * true_tokens / generated_tokens_cap
+                print_rank_0("=== overflow packing summary ===")
+                print_rank_0(f" > requested samples      : {requested_samples}")
+                print_rank_0(f" > generated samples      : {generated_samples}")
+                print_rank_0(f" > packing efficiency     : {packing_eff_pct:6.2f} % (non-pad tokens)")
+
                 np.save(doc_idx_filename, doc_idx, allow_pickle=True)
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: I guess all the mods from 306-358 are debug output? To remove?

File: megatron/data/gpt2_dataset.py [hunk 5]
```diff
@@ -385,14 +402,16 @@ def flush_bins(bins, doc_idx, sample_idx, sample_lengths):
                     cur_ptr = len(doc_idx)
                     cur_used = 0
                     for b in bins:
-                        if cur_used + b["used"] > TOKS_PER_SAMPLE:
-                            sample_idx.append([cur_ptr, 0])
-                            sample_lengths.append(cur_used)
+                        if cur_used + b["used"] >= TOKS_PER_SAMPLE:
+                            if cur_used > 0:
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: Unnecessary if statement kinda. It'll always come out to True

File: megatron/data/gpt2_dataset.py [hunk 6]
```diff
@@ -385,14 +402,16 @@ def flush_bins(bins, doc_idx, sample_idx, sample_lengths):
...
                         doc_idx.extend(b["docs"])
                         cur_used += b["used"]
-                    sample_idx.append([cur_ptr, 0])  # close final merged bin
-                    sample_lengths.append(cur_used)
+                    if cur_used > 0:
+                        sample_idx.append([cur_ptr, 0])  # close final merged bin
+                        sample_lengths.append(cur_used)
                     bins.clear()
 
                 # --------------------------------------------------------------
```
Comment (iPRET) [inline megatron/data/gpt2_dataset.py]: k_bin_packed seems legit ^_^

File: megatron/learning_rates.py [hunk 7]
```diff
@@ -96,14 +96,39 @@ def get_lr(self):
         elif self.decay_style == "cosine":
             end_iter_ = self.end_iter - self.warmup_iter
             lr = self.min_lr + (
-                (self.start_lr - self.min_lr)
-                / 2.0
-                * (math.cos(math.pi * num_iters_ / end_iter_) + 1)
+                    (self.start_lr - self.min_lr)
+                    / 2.0
+                    * (math.cos(math.pi * num_iters_ / end_iter_) + 1)
             )
```
Comment (iPRET) [inline megatron/learning_rates.py]: ty for fixing this indentation, I think the codebase is now better for it.

File: megatron/learning_rates.py [hunk 8]
```diff
@@ -96,14 +96,39 @@ def get_lr(self):
...
         elif self.decay_style == "exponential":
             # exp(-0.693) = 1/2
             end_iter = self.end_iter - self.warmup_iter
             lr = self.start_lr * math.exp(-0.693 * num_iters_ / end_iter)
+
+        elif self.decay_style == "sqrt":
+            # TODO: pass this as args
+            cd_start_iter = 194290
+            cd_end_iter = 194490
```
Comment (iPRET) [inline megatron/learning_rates.py]: Hardcoded incorrect iterations. Reconsider your life choices.

File: megatron/learning_rates.py [hunk 9]
```diff
@@ -96,14 +96,39 @@ def get_lr(self):
...
+            max_cd_lr = 1.6*10**(-4)
+            min_cd_lr = 1.6*10**(-6)
```
Comment (iPRET) [inline megatron/learning_rates.py]: hardcoded min and max lr. You sure you can't just reuse some of the already defined args for this?

File: megatron/learning_rates.py [hunk 10]
```diff
@@ -96,14 +96,39 @@ def get_lr(self):
...
+
+            if self.neox_args.iteration_offset:
+                global_num_iters_ = num_iters_ + self.neox_args.iteration_offset
+            else:
+                global_num_iters_ = num_iters_
+
+            if global_num_iters_ > cd_start_iter:
+
+                lr  = max_cd_lr - (max_cd_lr - min_cd_lr) * math.sqrt(
+                                (global_num_iters_ - cd_start_iter) / (cd_end_iter - cd_start_iter))
```
Comment (iPRET) [inline megatron/learning_rates.py]: Calculations seem legit otherwise ^_^~

File: megatron/training.py [hunk 11]
```diff
@@ -373,8 +373,36 @@ def _get_batch(neox_args, tokenizer, keys, data, datatype, label_mask_zero=False
...
+    # shifted_mask = torch.zeros_like(loss_mask)
+    # shifted_mask[:, :-1] = loss_mask[:, 1:]
+    # loss_mask = shifted_mask * label_mask
+
     # combine loss masks from get_ltor_masks_and_position_ids with loss masks from data
+    print_rank_0(f"label_mask : {label_mask}")
+    #print_cols_by_block(loss_mask, block=100, name="starting loss mask")
     loss_mask = label_mask.to(loss_mask.dtype) * loss_mask
+    #print_cols_by_block(loss_mask, block=100, name="final loss mask")
     return tokens, labels, loss_mask, attention_mask, position_ids
```
Comment (iPRET) [inline megatron/training.py]: Remove debug output plez

File: megatron/training.py [hunk 12]
```diff
@@ -618,6 +646,37 @@ def forward_step(
...
+                    print_rank_0(f"[{start:4d}:{end - 1:4d}] {vals}")
+
+
+        print_rank_0("Tokens: {}".format(tokens.size()))
+        batch_size, seq_length = tokens.size()
+        # for b in range(batch_size):
+        #     print_rank_0(f"Batch {b}, tokens: {tokens[b]}")
+        #print_cols_by_block(tokens, block=100, name="tokens")
+
         # Sequential returns moe_losses, but this is not yet supported by pipe parallel
```
Comment (iPRET) [inline megatron/training.py]: Remove debug output plz

File: megatron/utils.py [hunk 13]
```diff
@@ -119,11 +119,57 @@ def get_ltor_masks_and_position_ids(
         # convert to bool and flip
```
Comment (iPRET) [inline megatron/utils.py]: We had a more efficient implementation of attention mask construction I think.

File: megatron/utils.py [hunk 14]
```diff
@@ -119,11 +119,57 @@ def get_ltor_masks_and_position_ids(
...
+    # end-of-document masking
     if eod_mask_loss:
         loss_mask[data == eod_token] = 0.0
 
+    # # FIXME: remove?
+    # loss_mask[data == 8] = 0.0
+
+    # padding mask
+    loss_mask[data == 0] = 0.0
+
```
Comment (iPRET) [inline megatron/utils.py]: Didn't check for off by 1 errors, but otherwise seems legit >_<

File: misc/plot_data_distribution.py [hunk 15]
```diff
@@ -166,6 +197,10 @@ def create_separate_plot(categories, short, long, color, hatch_color, label_shor
     title4 = f"Total: {np.sum(total_short):.0f}B Short + {np.sum(total_long):.0f}B Long [{100 * np.sum(total_short) / total_sum:.2f}% Short, {100 * np.sum(total_long) / total_sum:.2f}% Long]"
     create_separate_plot(categories, total_short, total_long, 'peachpuff', '#D2691E', 'Total_Short', 'Total_Long', title4)
 
+    # Separate TOTAL Plot (All phases, Short and Long separated)
+    title4 = f"Total: {np.sum(total_short_and_e):.0f}B Short + {np.sum(total_long_and_e):.0f}B Long [{100 * np.sum(total_short_and_e) / total_sum_and_e:.2f}% Short, {100 * np.sum(total_long_and_e) / total_sum_and_e:.2f}% Long]  Σ: {total_sum_and_e:.0f} B"
+    create_separate_plot(categories, total_short_and_e, total_long_and_e, 'peachpuff', '#D2691E', 'Total_Short', 'Total_Long', title4)
+
```
Comment (iPRET) [inline misc/plot_data_distribution.py]: Random script I presume doesn't matter

File: tools/datasets/preprocess_sft.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        Encoder.tokenizer = build_tokenizer(self.args)
+
+    def encode(self, text):
+        if self.args.ftfy:
+            text = ftfy.fix_text(text)
+        ids = {}
+        for key in self.args.jsonl_keys:
+            doc_ids = []
+            print(f"text: {text}")
+            meme = text.split("<|mostuniqueseperatortokenintheworld|>")
```
Comment (iPRET) [inline tools/datasets/preprocess_sft.py]: Would be cool if the meme token could be cl-arg-ed.
Would also be nice if the structure wasn't assumed to be like Q - A
But any length of masked - unmasked - masked - unmasked - mask....

**Review by iPRET — CHANGES_REQUESTED** [review]
Not pullable

File: tools/datasets/preprocess_sft.py (hunk 16 above)
Comment (SilverSulfide) [reply tools/datasets/preprocess_sft.py]: yes, it would be cool : >
